### PR TITLE
Add paddle-server ROCK rocraft

### PR DESCRIPTION
**Pull request by misohu**
Steps to test 
```
cd paddle 
rockcraft clean && rockcraft pack --verbosity=trace
sudo skopeo --insecure-policy copy oci-archive:paddle-server_0.11.0_amd64.rock docker-daemon:charmedkubeflow/paddle-server:0.11.0
docker run charmedkubeflow/paddle-server:0.11.0
```

There should be an output with deprecation warnings like this 
```
2024-02-13T09:37:06.255Z [pebble] Started daemon.
2024-02-13T09:37:06.261Z [pebble] POST /v1/services 5.266914ms 202
2024-02-13T09:37:06.261Z [pebble] Started default services with change 1.
2024-02-13T09:37:06.265Z [pebble] Service "paddle-server" starting: python3.10 -m paddleserver [ args ]
2024-02-13T09:37:07.581Z [paddle-server] /usr/local/lib/python3.10/dist-packages/pkg_resources/__init__.py:121: DeprecationWarning: pkg_resources is deprecated as an API
2024-02-13T09:37:07.581Z [paddle-server]   warnings.warn("pkg_resources is deprecated as an API", DeprecationWarning)
2024-02-13T09:37:07.602Z [paddle-server] /usr/local/lib/python3.10/dist-packages/pkg_resources/__init__.py:2870: DeprecationWarning: Deprecated call to `pkg_resources.declare_namespace('google')`.
2024-02-13T09:37:07.602Z [paddle-server] Implementing implicit namespace packages (as specified in PEP 420) is preferred to `pkg_resources.declare_namespace`. See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-namespace-packages
2024-02-13T09:37:07.602Z [paddle-server]   declare_namespace(pkg)
2024-02-13T09:37:07.603Z [paddle-server] /usr/local/lib/python3.10/dist-packages/pkg_resources/__init__.py:2870: DeprecationWarning: Deprecated call to `pkg_resources.declare_namespace('google.cloud')`.
2024-02-13T09:37:07.603Z [paddle-server] Implementing implicit namespace packages (as specified in PEP 420) is preferred to `pkg_resources.declare_namespace`. See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-namespace-packages
2024-02-13T09:37:07.603Z [paddle-server]   declare_namespace(pkg)
2024-02-13T09:37:07.603Z [paddle-server] /usr/local/lib/python3.10/dist-packages/pkg_resources/__init__.py:2349: DeprecationWarning: Deprecated call to `pkg_resources.declare_namespace('google')`.
2024-02-13T09:37:07.603Z [paddle-server] Implementing implicit namespace packages (as specified in PEP 420) is preferred to `pkg_resources.declare_namespace`. See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-namespace-packages
2024-02-13T09:37:07.603Z [paddle-server]   declare_namespace(parent)
2024-02-13T09:37:07.603Z [paddle-server] /usr/local/lib/python3.10/dist-packages/pkg_resources/__init__.py:2870: DeprecationWarning: Deprecated call to `pkg_resources.declare_namespace('google.logging')`.
2024-02-13T09:37:07.603Z [paddle-server] Implementing implicit namespace packages (as specified in PEP 420) is preferred to `pkg_resources.declare_namespace`. See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-namespace-packages
2024-02-13T09:37:07.603Z [paddle-server]   declare_namespace(pkg)
2024-02-13T09:37:07.861Z [paddle-server] /usr/local/lib/python3.10/dist-packages/paddle/profiler/profiler_statistic.py:1402: DeprecationWarning: invalid escape sequence '\('
2024-02-13T09:37:07.861Z [paddle-server]   kernel_name_pattern = re.compile('(.+?)(<.*>)(\(.*\))')
2024-02-13T09:37:08.293Z [paddle-server] /usr/local/lib/python3.10/dist-packages/paddle/tensor/random.py:31: DeprecationWarning: invalid escape sequence '\e'
2024-02-13T09:37:08.293Z [paddle-server]   """
2024-02-13T09:37:08.482Z [paddle-server] /usr/local/lib/python3.10/dist-packages/paddle/distributed/parallel.py:87: SyntaxWarning: "is" with a literal. Did you mean "=="?
2024-02-13T09:37:08.482Z [paddle-server]   ) or backend is 'xccl':
2024-02-13T09:37:08.849Z [paddle-server] /usr/local/lib/python3.10/dist-packages/paddle/fluid/contrib/slim/quantization/post_training_quantization.py:438: SyntaxWarning: "is" with a literal. Did you mean "=="?
2024-02-13T09:37:08.849Z [paddle-server]   if self._algo is 'min_max':
2024-02-13T09:37:09.101Z [paddle-server] /usr/local/lib/python3.10/dist-packages/paddle/incubate/nn/layer/fused_linear.py:20: DeprecationWarning: invalid escape sequence '\_'
2024-02-13T09:37:09.101Z [paddle-server]   """
2024-02-13T09:37:09.164Z [paddle-server] /usr/local/lib/python3.10/dist-packages/paddle/distributed/fleet/meta_parallel/sharding/group_sharded_optimizer_stage2.py:208: DeprecationWarning: invalid escape sequence '\.'
2024-02-13T09:37:09.164Z [paddle-server]   "The param name passed to the optimizer doesn't follow .+_[0-9]+\..+ patter, "
2024-02-13T09:37:09.204Z [paddle-server] /usr/local/lib/python3.10/dist-packages/paddle/dataset/common.py:38: DeprecationWarning: invalid escape sequence '\T'
2024-02-13T09:37:09.204Z [paddle-server]   """
2024-02-13T09:37:09.357Z [paddle-server] /usr/local/lib/python3.10/dist-packages/paddle/sparse/unary.py:555: DeprecationWarning: invalid escape sequence '\p'
2024-02-13T09:37:09.357Z [paddle-server]   """
2024-02-13T09:37:09.357Z [paddle-server] /usr/local/lib/python3.10/dist-packages/paddle/sparse/unary.py:588: DeprecationWarning: invalid escape sequence '\p'
2024-02-13T09:37:09.357Z [paddle-server]   """
2024-02-13T09:37:09.362Z [paddle-server] /usr/local/lib/python3.10/dist-packages/paddle/sparse/nn/functional/transformer.py:29: DeprecationWarning: invalid escape sequence '\s'
2024-02-13T09:37:09.362Z [paddle-server]   """
2024-02-13T09:37:09.362Z [paddle-server] /usr/local/lib/python3.10/dist-packages/paddle/sparse/nn/functional/activation.py:64: DeprecationWarning: invalid escape sequence '\e'
2024-02-13T09:37:09.362Z [paddle-server]   """
2024-02-13T09:37:09.362Z [paddle-server] /usr/local/lib/python3.10/dist-packages/paddle/sparse/nn/functional/activation.py:149: DeprecationWarning: invalid escape sequence '\_'
2024-02-13T09:37:09.362Z [paddle-server]   """
2024-02-13T09:37:14.459Z [paddle-server] /usr/local/lib/python3.10/dist-packages/google/rpc/__init__.py:20: DeprecationWarning: Deprecated call to `pkg_resources.declare_namespace('google.rpc')`.
2024-02-13T09:37:14.459Z [paddle-server] Implementing implicit namespace packages (as specified in PEP 420) is preferred to `pkg_resources.declare_namespace`. See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-namespace-packages
2024-02-13T09:37:14.459Z [paddle-server]   pkg_resources.declare_namespace(__name__)
2024-02-13T09:37:14.459Z [paddle-server] /usr/local/lib/python3.10/dist-packages/pkg_resources/__init__.py:2349: DeprecationWarning: Deprecated call to `pkg_resources.declare_namespace('google')`.
2024-02-13T09:37:14.459Z [paddle-server] Implementing implicit namespace packages (as specified in PEP 420) is preferred to `pkg_resources.declare_namespace`. See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-namespace-packages
2024-02-13T09:37:14.459Z [paddle-server]   declare_namespace(parent)
2024-02-13T09:37:14.528Z [paddle-server] usage: __main__.py [-h] [--http_port HTTP_PORT] [--grpc_port GRPC_PORT]
2024-02-13T09:37:14.528Z [paddle-server]                    [--workers WORKERS] [--max_threads MAX_THREADS]
2024-02-13T09:37:14.528Z [paddle-server]                    [--max_asyncio_workers MAX_ASYNCIO_WORKERS]
2024-02-13T09:37:14.528Z [paddle-server]                    [--enable_grpc ENABLE_GRPC]
2024-02-13T09:37:14.528Z [paddle-server]                    [--enable_docs_url ENABLE_DOCS_URL]
2024-02-13T09:37:14.528Z [paddle-server]                    [--enable_latency_logging ENABLE_LATENCY_LOGGING]
2024-02-13T09:37:14.528Z [paddle-server]                    [--configure_logging CONFIGURE_LOGGING]
2024-02-13T09:37:14.528Z [paddle-server]                    [--log_config_file LOG_CONFIG_FILE]
2024-02-13T09:37:14.528Z [paddle-server]                    [--access_log_format ACCESS_LOG_FORMAT] --model_dir
2024-02-13T09:37:14.528Z [paddle-server]                    MODEL_DIR [--model_name MODEL_NAME]
2024-02-13T09:37:14.528Z [paddle-server] __main__.py: error: the following arguments are required: --model_dir
```

The output from workload service is identical to upstream container 
```
docker run kserve/paddleserver:v0.11.1
```

### CODE REVIEW

**Comment by kimwnasptd** [comment]
@misohu shouldn't we add also the CI automation in the repo for the ROCKs to get build/published on merge?

**Comment by ca-scribner** [comment]
I am +1 to adding the CI @kimwnasptd, but it should be as a separate PR

The settings on this repo are also not correct (I see there's no branch protection, for example).  I'll split it with you michal - you add the CI, I'll change the settings :D

**Comment by ca-scribner** [comment]
[this](https://github.com/canonical/filebrowser-rock/commit/f78008f746f7d7fa6d96d2bb560c7d8c02662de1) is an example of me adding the default rock CI elsewhere

then I used it in my [later pr](https://github.com/canonical/filebrowser-rock/pull/2) by adding the rock-specific files

File: paddle/rockcraft.yaml [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+base: ubuntu@22.04
+platforms:
+    amd64:
+run-user: _daemon_
+services:
+  paddle-server:
+    override: replace
+    summary: "Paddle server service"
+    startup: enabled
+    command: "python3.10 -m paddleserver [ ]"
```
Comment (ca-scribner) [inline paddle/rockcraft.yaml:16]: I feel like there's something we can do in staging the rock to make this work with just upstream's `python -m paddleserver` so we don't need to include the version

One way might just be making a symbolic link, but I wonder if there is anything better

File: paddle/rockcraft.yaml [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    plugin: nil
+    source: https://github.com/kserve/kserve.git
+    source-subdir: python
+    source-tag: v0.11.0
+    build-packages:
+    - python3.10
+    overlay-packages:
+    - python3.10
+    - build-essential
+    - libgomp1
```
[the remark this answers is not in the document]
Comment (ca-scribner) [reply paddle/rockcraft.yaml:40]: Oh wait, is this telling rockcraft to take the python3.10 from the build packages and stage it?  I'm not sure

**Comment by ca-scribner** [comment]
Looking at the parent issue #18, it does explicitly say publish manually.  But I'm with @kimwnasptd in that I think we should change it to be automatic

File: paddle/rockcraft.yaml
Comment (ca-scribner) [inline paddle/rockcraft.yaml:1]: Can we rename this directory to `paddleserver` to be consistent with the [upstream image's name](https://hub.docker.com/r/kserve/paddleserver/tags)

File: paddle/rockcraft.yaml [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+      dpkg-query --root=${CRAFT_PROJECT_DIR}/../bundles/ubuntu-22.04/rootfs/ -f '${db:Status-Abbrev},${binary:Package},${Version},${source:Package},${Source:Version}\n' -W) \
+      > ${CRAFT_PART_INSTALL}/usr/share/rocks/dpkg.query
+
+  python:
+    plugin: nil
+    source: https://github.com/kserve/kserve.git
+    source-subdir: python
+    source-tag: v0.11.0
+    build-packages:
+    - python3.10
```
Comment (ca-scribner) [inline paddle/rockcraft.yaml:36]: Why do we use `build-packages` here and not `stage-packages`?  I'm still digesting, but I think what we are doing here is:
* putting python3.10 into the part for building step only
* during build step, installing packages that end up in the `/usr/local/lib/python3.10/dist-packages`
* priming the `/usr/local/lib/python3.10/dist-packages` **but not the python3.10 we used to construct that dir**
* in the final image, using the final image's native python3.10 which is accessing the above mentioned `/usr/local/lib/python3.10/dist-packages`

is that how this works?

File: paddle/rockcraft.yaml (hunk 1 above)
Comment (misohu) [reply paddle/rockcraft.yaml:16]: I dont think this adds any value tbh :) lets skip

File: paddle/rockcraft.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (misohu) [reply paddle/rockcraft.yaml:40]: I understand it that it wants to use it in both build and install

File: paddle/rockcraft.yaml (hunk 3 above)
Comment (misohu) [reply paddle/rockcraft.yaml:36]: yes fresh python3.10 ... I remember rockcraft team recommending that with pmml server to get rid of unnecesery packages from build e.g. poetry.

Comment (misohu) [reply paddle/rockcraft.yaml]: Yes I think so

**Comment by misohu** [comment]
@ca-scribner @kimwnasptd  here is PR for CI https://github.com/canonical/kserve-rocks/pull/22

File: paddle/rockcraft.yaml (hunk 3 above)
Comment (misohu) [reply paddle/rockcraft.yaml]: Removed

**Comment by ca-scribner** [comment]
I'll give this another review, but I'm still messing around with things in my related issue #19.  I think one or two things here might be redundant (maybe rockcraft changed since last we did this?)

File: paddleserver/dummy_pyproject.toml [hunk 4]
```diff
@@ -0,0 +1,11 @@
+[tool.poetry]
+name = "workaround-for-editable-install"
+version = "0.0.1"
+description = ""
+authors = ["none"]
+
+[tool.poetry.dependencies]
+python = ">=3.8,<3.11"
```
Comment (DnPlas) [inline paddleserver/dummy_pyproject.toml:8]: Should we add a small comment here that the upper version of python should be +1 minor version of the one we install in the overlay ?

**Review by DnPlas — COMMENTED** [review]
Everything looks good, just a super small comment. Thanks!

File: paddleserver/dummy_pyproject.toml (hunk 4 above)
Comment (ca-scribner) [reply paddleserver/dummy_pyproject.toml:8]: These come from the kserve/paddle packages we're installing from upstream (see [here](https://github.com/kserve/kserve/blob/v0.11.2/python/paddleserver/pyproject.toml#L13)).  Would a comment here help clarify that?  

For reasons that aren't clear to me, upstream's packages have different ranges (this is <3.12, paddle is <3.11, etc) so I've matched their ranges here.

**Review by DnPlas — APPROVED** [review]

**Pull request merged**